actionpack/lib/action_dispatch/middleware/server_timing.rb

```diff
@@ -0,0 +1,77 @@
+# frozen_string_literal: true
+
+module ActionDispatch
+  class ServerTiming
+    class << self
+      def add_timing(key:, dur: nil, desc: nil)
+        timings[key] = { dur: dur, desc: desc }
+      end
+
+      def to_header
+        headers = []
+        timings.each_pair do |key, data|
+          header = "#{key}"
+          header += ";desc=\"#{data[:desc]}\"" if data[:desc]
+          header += ";dur=#{data[:dur]}" if data[:dur]
+          headers << header
+        end
+        headers.join(", ")
+      end
+
+      def measure(key:, desc: nil, &blk)
+        ActiveSupport::Notifications.instrument("measure.server_timing", desc: desc, key: key) do
+          yield
+        end
+      end
+
+      def clear_timings!
+        Thread.current[:server_timing_registry] = {}
+      end
+
+      def timings
+        Thread.current[:server_timing_registry] ||= {}
```

Comment:
If this middleware is going to be below the `ActionDispatch::Executor` one in the middleware stack, we can probably piggyback on top of AS::CurrentAttributes.

Reply:
I like this a lot, and I'm gonna take this PR in that direction as I refine it. 

Though I was thinking of a way to support doing something like: 

``` ruby
threads =[]
files.each do |file|
  Thread.new do
    ActionDispatch::ServerTiming.measure(key: :file_download, desc: "download") do
      file.download!
    end
  end
end
threads.join
```

unfortunately `AS::CurrentAttributes` doesn't give us a benefit here since it's still in a thread local .